cmd/snap/cmd_userd.go

```diff
@@ -72,3 +85,117 @@ func (x *cmdUserd) Execute(args []string) error {
 
 	return x.userd.Stop()
 }
+
+var (
+	replacedDesktopKeys = []string{"%f", "%F", "%u", "%U", "%d", "%D",
+		"%n", "%N", "%i", "%c", "%k", "%v", "%m"}
+)
+
+func findExec(desktopFilePath string) (string, error) {
+	f, err := os.Open(desktopFilePath)
+	if err != nil {
+		return "", err
+	}
+	defer f.Close()
+
+	scanner := bufio.NewScanner(f)
+	execCmd := ""
+	for scanner.Scan() {
+		bline := scanner.Bytes()
+
+		if !bytes.HasPrefix(bline, []byte("Exec=")) {
+			continue
+		}
+
+		execCmd = strings.Split(string(bline), "Exec=")[1]
+		for _, key := range replacedDesktopKeys {
+			execCmd = strings.Replace(execCmd, key, "", -1)
+		}
+		return execCmd, nil
+	}
```

Comment:
I was expecting a small refactor of wrappers/desktop.go `sanitizeDesktopFile()` so you could call it, it would verify the file, rewrite the Exec line and you could just use the rewritten Exec line. This was we have validation and rewriting in one spot. Is there a reason you didn't do this?

Reply:
I believe we opted for the simpler solution of just looking at the `Exec=` line and ignoring the rest. Not rewriting the file saves us the hassle of keeping the updated copies around. Even if we were to patch the file in place, we'd still need to parse the file and verify the that the Exec is correct.

Reply:
I wasn't proposing rewriting the file. I was saying refactor sanitizeDesktopFile() such that the validation checks it makes can be called by your code such that you get back the contents of the rewritten file in memory. In addition to the Exec= line rewrite happening in one place in the code (rather than two, like this PR would do), you get validation of the entire file for free as a safeguard for a potential future where this code becomes interested in other fields.

Reply:
I've split it up for a simpler review in #4781